### PR TITLE
(TWILL-186) Fix NPE and container size mismatch

### PR DESCRIPTION
**Pull request by chtyim**
The code changes are separated into two commits for easy review.

1. The first commit is refactoring/cleanup to make AM code cleaner and easier to read. Also fix a potential tight loop issue in the `doRun` method during container restart
2. The second commit is the actual fix for the TWILL-186 issue.

### CODE REVIEW

File: twill-yarn/src/main/java/org/apache/twill/internal/yarn/AbstractYarnAMClient.java [hunk 1]
```diff
@@ -50,12 +51,11 @@
   private static final Logger LOG = LoggerFactory.getLogger(AbstractYarnAMClient.class);
 
   // Map from a unique ID to inflight requests
-  private final Multimap<String, T> containerRequests;
-
-  // List of requests pending to send through allocate call
-  private final List<T> requests;
+  private final Multimap<String, T> inflightRequests;
+  // Map from a unique ID to pending requests. It is for recording
```
Comment (anwar6953) [inline twill-yarn/src/main/java/org/apache/twill/internal/yarn/AbstractYarnAMClient.java]: It is for recording what?
(incomplete sentence?)

Comment (chtyim) [inline twill-yarn/src/main/java/org/apache/twill/internal/yarn/AbstractYarnAMClient.java]: Oh. It is for recording the container requests that has yet to be sent to RM. Will update the comment.

**Comment by anwar6953** [comment]
LGTM

**Comment by hsaputra** [comment]
Hi @chtyim looks like there is only 1 commit for this PR?

**Comment by chtyim** [comment]
I squashed them after getting LGTM to prepare for the merge

**Comment by chtyim** [comment]
Seems like the github sync is lagging. I merged this change about 6 hours ago

**Comment by hsaputra** [comment]
Ah ok

**Pull request closed without merging**